src/pages/Swap/LimitOrderBox/components/LimitOrderForm/LimitOrderForm.tsx

```diff
@@ -128,34 +127,37 @@ export function LimitOrderForm({ account, provider, chainId }: LimitOrderFormPro
         limitOrder.kind === LimitOrderKind.SELL ? sellTokenAmount.currency.decimals : buyTokenAmount.currency.decimals
       ).toString()
 
-      // get and parse the sell token amount
-      const sellTokenAmountFloat = parseFloat(
-        formatUnits(sellTokenAmount.raw.toString(), sellTokenAmount.currency.decimals)
-      )
-
-      let newBuyAmountAsFloat = 0 // the amount of buy token
-
-      if (limitOrder.kind === LimitOrderKind.SELL) {
-        newBuyAmountAsFloat = sellTokenAmountFloat * nextLimitPriceFloat
+      if (inputFocus === InputFocus.SELL) {
+        const {
+          amount: newBuyAmountFloat,
+          buyAmountWei: nextBuyAmountWei,
+          newBuyTokenAmount,
+        } = computeNewAmount(buyTokenAmount, sellTokenAmount, nextLimitPriceFloat, limitOrder.kind, inputFocus)
+
+        setBuyTokenAmount(newBuyTokenAmount)
+        setFormattedLimitPrice(toFixedSix(nextLimitPriceFloat))
+        setFormattedBuyAmount(toFixedSix(newBuyAmountFloat))
+        setLimitOrder({
+          ...limitOrder,
+          limitPrice: limitPrice,
+          buyAmount: nextBuyAmountWei,
+        })
       } else {
-        newBuyAmountAsFloat = sellTokenAmountFloat / nextLimitPriceFloat
+        const {
+          amount: newSellAmountFloat,
+          sellAmountWei: nextSellAmountWei,
+          newSellTokenAmount,
+        } = computeNewAmount(buyTokenAmount, sellTokenAmount, nextLimitPriceFloat, limitOrder.kind, inputFocus)
+
+        setSellTokenAmount(newSellTokenAmount)
+        setFormattedLimitPrice(toFixedSix(nextLimitPriceFloat))
+        setFormattedSellAmount(toFixedSix(newSellAmountFloat))
+        setLimitOrder({
+          ...limitOrder,
+          limitPrice: limitPrice,
+          sellAmount: nextSellAmountWei,
+        })
```

Comment:
We should consider refactoring this to `useReducer` approach in the future.

Reply:
100% agree, it will be better for state updates